### ISSUE TITLE
Socket-IO Flask-Login - current_user still logged in after logging out

### ISSUE THREAD

**Issue by KemperVreden** [miguelgrinberg/Flask-SocketIO#278]
Hello Miguel

In a Flask-based Web-Application I am using Flask-SocketIO and Flask-Login. The Application is served by the eventlet Network Library using the wrapper of Flask-SocketIO.

To add basic user authentication I have implemented the Flask-Login extension. This is working as expected. All those Routes which need User Authentication are being decorated by the Flask-Login decorator, just like the following example:

    @blueprint.route('/')
    @flask_login.login_required
    def index():
        return render_template('point-measurement.html')

In addition I need Flask-SocketIO to send updates to the client. When the user requests such a page, a socket-io connection is established from that page to the Web-Application. This works as expected.

But I also wanted to add authentication to the Socket-IO handlers and read about the Flask-Login integration in the chapter of the Flask-SocketIO documentation. Next I implemented a decorator like that in the documentation and added it to the corresponding socket-handler:

    def authenticated_only(f):
        @functools.wraps(f)
        def wrapped(*args, **kwargs):
            if not current_user.is_authenticated:
                disconnect()
            else:                
                return f(*args, **kwargs)
        return wrapped

And here's an example of how I added the decorator to a socket-io handler:

    @socket_io.on('start_measurement', namespace=SOCKET_IO_NAMESPACE) 
    @authenticated_only
    def start_measurement():
        print "Start Measurement"

Ok. Now comes the strange part: When I log out the current user from the web-application - I am doing this with calling a specific URL (e.g. http://server.address/logout) - The user is being logged out.

As expected I cannot access those URL Routes that are decorated by the Flask-Login decorator ('`login_required`')

But when I log-out the current user while the client is still being connected to the web-application (I am calling the log-out URL from another Tab in the same Browser), the next time the client sends a message over that socket, the Socket-IO Decorator says that the `current_user` is still logged in. 

By checking the same with the standard Flask-Login Decorator after calling one of the standard URL-Routes, that decorator says that `current_user == anonymous_user`. This is as expected.

The Socket-IO Authorization-Decorator holds the `current_user` as still logged in until I restart the whole web-application.

I am just trying to figure out what I got wrong here but I can't find an answer.

**Comment by KemperVreden** [comment]
While being still working on this I have found out that calling current_user on the authorization decorator for Flask url routes is returning the current user as expected - so when logged in it returns the UserMixin-Instance / when logged out it returns an instance of AnonymousUserMixin.

The authorization decorator for Flask-SocketIO Handler always returns a UserMixin-Instance after the user has logged in - even after logging out. Only a restart of the web-application leads to the right current_user type. 

I have also found out that the Socket-IO Handler authorization handler ALWAYS calls user_loader decorator / callback.

**Comment by KemperVreden** [comment]
In your documentation I just read:

> A request context is also pushed before invoking a handler, also making request and session available. But note that WebSocket events do not have individual requests associated with them, so the request context that started the connection is pushed for all the events that are dispatched during the life of the connection.

Does that mean that current_user references to the request context that started the connection since this is being pushed to all future events? 

If this is the case then it makes sense why current_user still shows the last logged in user - even after logging out - and that only a restart of the web app brings back wanted behavior ...

**Comment by KemperVreden** [comment]
Ok. It seems like that was the issue ... But I am leaving this open since I am waiting for your answer .

**Comment by KemperVreden** [comment]
BTW. I understand that at the time the client is connected with the server and the socket connection is fully established, there are no request, etc. 

But I think that when in the meantime the user logs out, the current_user has to change so that one can do check ups every time a new message is being received / handled through the socket. This has to be possible to implement, or?

**Comment by miguelgrinberg** [comment]
Well, looks like you figured out the problem by yourself. You are correct, when the client connects to the Socket.IO server, a snapshot of the user session is made available to that connection. If the user session changes from the HTTP side (as in when the user logs out), those changes cannot be made easily available on the socket side, because the socket connection is in the context of the request that started it, which happened long ago.

Really the limitation here is that the user session is in a cookie, and WebSocket does not have cookies. If you really need to make changes from the HTTP side visible on the WebSocket side, all you need to do is switch to server-side sessions, as indicated in the documentation. If your sessions are stored server-side then they can be accessed and modified by both http and websocket.

**Comment by KemperVreden** [comment]
@miguelgrinberg Thanks for the advice! I have integrated Server-Side Sessions and the authorization decorator now works as expected. I am not sure however if this is really necessary for my web-application but it helped me get a better understanding of web-applications, request and context in general. 

BTW. Is there something against using Server-Side Sessions? I've integrated the Flask-Session extension and used the filesystem Session type.
